keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,200 @@
+      * [Motivation](#motivation)
+      * [User Experience](#user-experience)
+         * [Use Cases](#use-cases)
+      * [Proposal](#proposal)
+      * [Implementation](#implementation)
+         * [API change](#api-change)
+         * [Networked volume plugins](#networked-volume-plugins)
+         * [Local storage](#local-storage)
+         * [PV controller changes](#pv-controller-changes)
+      * [Implementation History](#implementation-history)
+
+
+## Motivation
+
+For now, kubernetes has no way to monitor PVs, which may cause serious problems. 
+
+For example: if volumes are unhealthy, and pods still try to write data to them, which will lead to data loss. 
+
+And also if nodes break down, local PVs in the nodes therefore can not be accessed any more. 
+
+So it is necessary to have a mechanism for monitoring PVs.
+
+## User Experience
+
+### User Cases
+
+- If the local PV directory is deleted, users should know that and the local PV should be tainted;
+
+- If the local PV path is not a mountpoint any more, the local PV should be tainted;
+
+- If the node with local PVs in it breaks down, the local PVs should be tainted;
+
+- If the volume provisioned by the driver is deleted, the PV object in kubernetes should be tainted;
+
+- If we can not get access to the PV volume for a certain time (network or some other problems), we need to taint the PV;
+
+- PV fsType checking ? bad blocks checking ?
+
+- If an attached volume is no longer attached, the PV object should be tainted
+
+- If a mounted volume is not longer mounted, the PV object should be tainted.
+
+
+## Proposal
+
+In this proposal, we only focus on PV monitoring. 
+
+API change, local PVs and network PVs monitor are the three main parts we want to discuss here.
+
+Regarding the API part, we don’t plan to change PV API object at the first stage, instead we may use the VolumeTaint introduced in this proposal . We propose to create three taints for PV monitoring, details can be found in Implementation section below.
+
+For local PVs, we need to create an **agent** deployed to each node in the cluster to check the local PVs health condition and taint them if necessary.  A local PV monitor **controller** is also needed responsible for watching node failure events and tainting local PVs if nodes break down. BTW: we haven’t support local volumes in CSI, so CSI change is not needed here. 
+
+For networked storage, we just need to create a **controller** to check the volumes’ status periodically, and each volume plugin should implement its own health checking methods. 
+Since we are migrating in-tree volume plugins out of tree, the health checking methods should be supported in CSI spec. Detailed CSI spec info is [here](https://docs.google.com/document/d/1mjYlgXflAayLMPVFNT-qf3iJTVJYJBD-2RQh1jZTJ6s/edit#heading=h.3uxdz6kyha0s).
+
+The whole architecture is as below:
+![pv health monitor architecture](./pv-health-monitor.png)
+
+
+## Implementation
+
```

Comment:
in the case of a reserved volume: shared between multiple nodes, but RWO and this node does not own the reservation, do we need such a state? The volume could be functioning perfectly well, but is not available to this node at the moment.

Reply:
Good question, if the volume itself is problematical, we should definitely taint it.
If the volume is shared between multiple nodes and RWX, and because of network problem or sth like that, it is not available to one specific node, i think we need to taint it, it is an attaching problem.
But if it is RWO, and the node does not own the reservation, IIUC, the volume is healthy and will not lead to data loss, as you mentioned above, it may lead to reduced performance.I prefer not tainting it at first.
It seems that we need to list the specific causes. I will rethink the extension mechanism, and validate if Taint is suitable here.